src/string.rs

```diff
@@ -0,0 +1,194 @@
+            BindingOptions::ValidBindings((0..data.len()).map_into().collect())
+        } else {
+            let Some(StringPosition(root_pos)) = known_bindings.get(&Self::Key::root()) else {
+                return BindingOptions::MissingIndexKeys(vec![Self::Key::root()]);
+            };
+
+            let &StringIndexKey(offset) = key;
+
+            BindingOptions::ValidBindings(vec![(root_pos + offset).into()])
+        }
+    }
+}
+
+/// A position in a string.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, From, Into)]
+pub struct StringPosition(usize);
+
+/// An index key, obtained from the position index of the character in the pattern.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, From, Into)]
+pub struct StringIndexKey(usize);
+
+impl StringIndexKey {
+    fn root() -> Self {
+        Self(0)
+    }
+}
+
+impl Debug for StringIndexKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "char@{}", self.0)
+    }
+}
+
+#[cfg(test)]
+pub(super) mod tests {
+    use crate::HashMap;
+
+    use rstest::rstest;
+
+    use self::pattern::StringPattern;
+    use crate::{ManyMatcher, PatternMatch, PortMatcher};
+
+    use super::*;
+
+    fn non_det_matcher(patterns: Vec<StringPattern>) -> StringManyMatcher {
+        ManyMatcher::from_patterns_with_det_heuristic(patterns, |_| false)
+    }
+
+    fn default_matcher(patterns: Vec<StringPattern>) -> StringManyMatcher {
+        ManyMatcher::from_patterns(patterns)
+    }
+
+    fn det_matcher(patterns: Vec<StringPattern>) -> StringManyMatcher {
+        ManyMatcher::from_patterns_with_det_heuristic(patterns, |_| true)
+    }
+
+    pub(super) const MATCHER_FACTORIES: &[fn(Vec<StringPattern>) -> StringManyMatcher] =
+        &[non_det_matcher, default_matcher, det_matcher];
+
+    /// Comparing two lists of pattern matches.
+    ///
+    /// We cannot use equality for two reasons:
+    ///  - we ignore the ordering of the matches
+    ///  - not all keys may be bound depending on the order things were matched
+    ///
+    /// The simplest way to identify a match is with the pair (PatternID, start_pos).
+    pub(super) fn pattern_match_eq(
+        expected: &[PatternMatch<HashMap<StringIndexKey, StringPosition>>],
+        actual: &[PatternMatch<HashMap<StringIndexKey, StringPosition>>],
+    ) -> bool {
+        let get_match_key =
+            |PatternMatch {
+                 pattern,
+                 match_data,
+             }: &PatternMatch<HashMap<StringIndexKey, StringPosition>>| {
+                let Some((StringIndexKey(offset), StringPosition(pos))) = match_data.iter().next()
+                else {
+                    panic!("Empty match data");
+                };
+                (*pattern, pos - offset)
+            };
+        // Build a map from pattern ID to all positions where it matches
+        let get_slice_key = |matches: &[PatternMatch<_>]| {
```

Comment:
optional/nit: but, consider
* making `get_match_key` operate only on the HashMap (caller passes the `pattern` into the tuple)
* combine the `map` and `fold` into just `fold` (over a list of `PatternMatch<_>`, calling the (new) `get_match_key`). `fold` might be better done as `for`...
* rename `get_slice_key` to `positions_per_pattern`

Reply:
has been simplified away too :)